py/formatfloat: Format ALL whole-number floats without decimal cruft.

dpwe opened the pull request:
Formerly, py/formatfloat would print whole numbers inaccurately
with nonzero digits beyond the decimal place.  This resulted
from its strategy of successive scaling of the argument by 0.1
which cannot be exactly represented in floating point.  This
change avoids scaling until the value is smaller than 1, so
all whole numbers print with zero fractional part.

This is to fix issue #4212.

This PR is an alternative to #8901.  It uses a different algorithm, using
floating-point arithmetic instead of uint32_t.  This makes it able to
work for all whole values (not only those smaller than 2^32).  It may
also reduce code size (to be verified).

Signed-off-by: Dan Ellis dan.ellis@gmail.com

dpwe committed:
py/formatfloat: Format all whole-number floats exactly.

Formerly, py/formatfloat would print whole numbers inaccurately
with nonzero digits beyond the decimal place.  This resulted
from its strategy of successive scaling of the argument by 0.1
which cannot be exactly represented in floating point.  This
change avoids scaling until the value is smaller than 1, so
all whole numbers print with zero fractional part.

This is to fix issue #4212

Signed-off-by: Dan Ellis dan.ellis@gmail.com

dpwe commented:
I'm happy with the code here, but it's failing the float_parse_doubleprec test for unix port / nanbox.  Any advice on how I can debug this?

Also, I'd be interested to see the code size comparisons.

dpgeorge commented:
This is the size-diff I get for this PR:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:  +176 +0.034% 
unix nanbox:  +144 +0.031% 
      stm32:  +156 +0.039% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:  +256 +0.037% GENERIC
      esp32:  +464 +0.031% GENERIC
     mimxrt:  +144 +0.040% TEENSY40
 renesas-ra:  +152 +0.024% RA6M2_EK
        nrf:  +144 +0.077% pca10040
        rp2:  +160 +0.031% PICO
       samd:  +156 +0.111% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

dpwe commented:
Thanks for the help.  Unfortunately, VARIANT=nanbox doesn't link on my dev system (MacOS 10.15.17, "warning: The i386 architecture is deprecated for macOS") although I can try it on my linode.

Do you happen to know what is different about nanbox?  It's as if it's failing some fairly specific floating-point comparisons.  I tried Googling nanbox, and I understand the concept of nan-boxing, but I'm not clear how it's used in MicroPython, or why it would affect non-NaN floating point operations.

dpwe commented:
Hmm, I was so sure I had fixed it!

It turns out the "difference" may lie partially in the parsing code rather than the formatting code.  On the plain build on my Mac, ```float("9" * 400 + "e-200")``` results in the float64 ```0x6974e718d7d7625c``` but on the nanbox build it is ```0x6974e718d7d7625a``` (2 eps smaller).  In fact, the nanbox value is closer to the specified value.

```
0x6974e718d7d7625a -> 99999999999999996973312221251036165947450327545502362648241750950346848435554075534196338404706251868027512415973882408182135734368278484639385041047239877871023591066789981811181813306167128854888448
0x6974e718d7d7625c -> 100000000000000030966143761524130482081095546903494511742201285480389933199978920360024169499249787174427657390648165217099222854144343466820462650820503200080302232128380506216383146771333146885488640
```
However, it's also true that the test case you sent (corresponding to the nanbox value, ```...625a```) renders differently on the two builds.  This appears to be because, on nanbox but not on plain unix, the expression ```(f >= next_u_base)``` evaluates to false if both ```f``` and ```next_u_base``` are float64s equal to that value -- so they should compare equal, but don't.  (I see it for several powers of 10, e.g. 1e23 ```0x4950912855330343670```.  But in that case, the rounding saves us).  Possibly worth raising a bug with the nanbox FP routines?

Overall, I don't think the test failure is a "bug" per se, it's just an imprecision.  And the fact that it passed under the old code was .. fluke?

If we changed the print in the test to use 15 sig figs, I believe it would work.  But asking you to change the test to pass my code feels .. problematic.

dpgeorge commented:
> This appears to be because, on nanbox but not on plain unix, the expression `(f >= next_u_base)` evaluates to false if both `f` and `next_u_base` are float64s equal to that value -- so they should compare equal, but don't.

Yes I'm also seeing this behaviour, that two values which have the same binary representation are not comparing equal.

I think the issue is that on x86 the internal precision of double operations is 80 bits.  And so even though they are equivalent 64-bit doubles, their 80-bit representations differ.

I could fix it by forcing a truncation of the 80-bit number:
```diff
@@ -265,9 +270,10 @@ int mp_format_float(FPTYPE f, char *buf, size_t buf_size, char fmt, int prec, ch
         // mantissa.
         FPTYPE u_base = FPCONST(1.0);
         for (e = 0, e1 = FPDECEXP; e1; e1 >>= 1, pos_pow++) {
-            if (f >= (u_base * *pos_pow)) {
+            volatile double xx = u_base * *pos_pow;
+            if (f >= xx) {
                 e += e1;
-                u_base *= *pos_pow;
+                u_base = xx;
             }
         }
```

That's not very satisfactory.

Can you change the `>=` in that comparison to `>`?  Then it won't rely on any edge behaviour.  Then you may need to adjust other bits of code to compensate for that change.

Alternatively, go back to the integer version #8901.

dpwe commented:
I tried testing for the equality case separately, after the initial loop, by subtracting the new base 10 times and seeing if the result was >= 0, but it failed the same way - again, presumably, because the compiler is optimizing by keeping the intermediate result in an 80 bit register.

Given that we're now coding around deviations from the ideal model, I think "volatile" is as good a solution as any - at least the reader knows that it's not about details of the math, but something in the compilation.

It's so interesting that this only came up for nanbox.  I assume it's non-IEEE-compliant to do comparisons without truncating down to 64 bits; maybe the deal with nanbox is that it's already noncompliant, so they pull in other optimizations too?

dpwe commented:
I got a little deeper into the code and simplified it, looking for code size reductions.

dpgeorge commented:
> It's so interesting that this only came up for nanbox. I assume it's non-IEEE-compliant to do comparisons without truncating down to 64 bits; maybe the deal with nanbox is that it's already noncompliant, so they pull in other optimizations too?

Nan-boxing mode is defined at our (MicroPython) end, it's a 32-bit x86 build with 64-bit objects.

The same bug appears in normal 32-bit builds of unix MicroPython.  The reason the CI didn't see it (which does build and test 32-bit unix MicroPython) is because the parsing of numbers is slightly different in nanbox vs non-nanbox (as noted up above).

You can see the error on a standard 32-bit build by using the test that gives the explicit representation of the double:
```python
import array
print(array.array('d', b'Zb\xd7\xd7\x18\xe7ti'))
print(array.array('d', b'}\xc3\x94%\xadI\xb2T'))
```
We should definitely add this exact test to the test suite.

If we go the route using `volatile`, it's only needed on x86 in 32-bit build mode, so can be guarded with `#if ...`.

dpgeorge commented:
Current code size diff is now:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:    +0 +0.000% 
unix nanbox:    +0 +0.000% 
      stm32:   +36 +0.009% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:  +104 +0.015% GENERIC
      esp32:  -196 -0.013% GENERIC
     mimxrt:   +16 +0.004% TEENSY40
 renesas-ra:   +32 +0.005% RA6M2_EK
        nrf:   +24 +0.013% pca10040
        rp2:   +24 +0.005% PICO
       samd:   +36 +0.026% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
That's surprising that unix is exactly 0... I double checked it and it seems to be true!

dpgeorge commented on py/formatfloat.c:
```diff
@@ -256,22 +244,26 @@ int mp_format_float(FPTYPE f, char *buf, size_t buf_size, char fmt, int prec, ch
+            // Use "volatile" to force the result to be stored in a float64.
+            // Without this, nanbox build would sometimes fail to report
+            // "equality" leading to weird results.
+            volatile FPTYPE next_u = u_base * *pos_pow;
```
Please guard the `volatile` with `#if defined(__i386__)`

dpgeorge commented on tests/float/float_format_ints.py:
```diff
@@ -0,0 +1,23 @@
+# represented by a float32 and that will also fit within a (signed) int32.
+# The upper bound of our integer-handling code is actually double this,
+# but that constant might cause trouble on systems using 32 bit ints.
+print("{:f}".format(2147483520))
```
Please add the 2 test cases using the explicit representation and array.array

dpwe replied:
Won't that fail on builds with
```
#define MICROPY_FLOAT_IMPL          (MICROPY_FLOAT_IMPL_FLOAT)
```
I guess there's some way to gate or condition tests based on build conditions.  I'll look into it.

dpwe replied:
I'm going with a separate ```tests/float/float_format_int_doubleprec.py```, which I'll add to the ```if upy_float_precision < 64``` skips in ```run-tests.py```.

dpgeorge replied:
Sounds good!

dpwe commented:
This looks good, right?

I regret to say there are still some integer powers of ten that don't work out perfectly, e.g.
```
>>> print(float(1e33))
9.999999999999998e+32
```
(33 is the smallest, and there are only 10 below 1e100).  I'm not sure how this is coming about, I guess because there are differences in how the parsing constructs the float, and how we build it in the formatting code (by multiplying-up factors of 10**(2**i)).

The unmodified code does this too, with the lowest being at 1e23.  However, there are only a total of 7 problem cases below 1e100.

dpwe commented:
To illustrate the 'aggressive rounding up' problem more completely, consider the following code fragment:

```
>>> d = int(1e16)
>>> for i in range(-10, 10):
...     print(d + i, "{:f}".format(float(d + i)))
... 
9999999999999990 9999999999999990.000000
9999999999999991 9999999999999992.000000
9999999999999992 9999999999999992.000000
9999999999999993 9999999999999992.000000
9999999999999994 9999999999999994.000000
9999999999999995 10000000000000000.000000
9999999999999996 10000000000000000.000000
9999999999999997 10000000000000000.000000
9999999999999998 10000000000000000.000000
9999999999999999 10000000000000000.000000
10000000000000000 10000000000000000.000000
10000000000000001 10000000000000000.000000
10000000000000002 10000000000000002.000000
10000000000000003 10000000000000004.000000
10000000000000004 10000000000000004.000000
10000000000000005 10000000000000004.000000
10000000000000006 10000000000000006.000000
10000000000000007 10000000000000008.000000
10000000000000008 10000000000000008.000000
10000000000000009 10000000000000008.000000
```
Notice the jump from ```...994.000...``` to ```...000.000...```.  We could have hoped for 996 and 998 in there.

However, compare the output from the current ```micropython:master``` code:

```
9999999999999990 9999999999999992.894573
9999999999999991 9999999999999992.894573
9999999999999992 9999999999999992.894573
9999999999999993 9999999999999992.894573
9999999999999994 9999999999999994.670929
9999999999999995 9999999999999998.223643
9999999999999996 9999999999999998.223643
9999999999999997 9999999999999998.223643
9999999999999998 10000000000000000.000000
9999999999999999 10000000000000000.000000
10000000000000000 10000000000000000.000000
10000000000000001 10000000000000000.000000
10000000000000002 10000000000000002.220446
10000000000000003 10000000000000004.440892
10000000000000004 10000000000000004.440892
10000000000000005 10000000000000004.440892
10000000000000006 10000000000000006.661338
10000000000000007 10000000000000008.881784
10000000000000008 10000000000000008.881784
10000000000000009 10000000000000008.881784
```
It manages to nail the 10^16 value (as it happens, although it doesn't for 10^23, 10^25, or 10^29), but all those other non-integer values - ugh.

dpwe commented:
Any next steps needed from me here?

dpgeorge commented:
I still need to check the code-size diff again, and review again.

Comparing this current version to #8901, how well did #8901 perform with this test you posted above:
```python
>>> d = int(1e16)
>>> for i in range(-10, 10):
...     print(d + i, "{:f}".format(float(d + i)))
```
?

dpwe commented:
Because 1e16 is outside the 4294967296.0 ~ 4e9 limit for the modified code path in #8901, this example renders the same under #8901 as for master:
```
>>> d = int(1e16)
>>> for i in range(-10, 10):
...     print(d + i, "{:f}".format(float(d + i)))
... 
9999999999999990 9999999999999992.894573
9999999999999991 9999999999999992.894573
9999999999999992 9999999999999992.894573
9999999999999993 9999999999999992.894573
9999999999999994 9999999999999994.670929
9999999999999995 9999999999999998.223643
9999999999999996 9999999999999998.223643
9999999999999997 9999999999999998.223643
9999999999999998 10000000000000000.000000
9999999999999999 10000000000000000.000000
10000000000000000 10000000000000000.000000
10000000000000001 10000000000000000.000000
10000000000000002 10000000000000002.220446
10000000000000003 10000000000000004.440892
10000000000000004 10000000000000004.440892
10000000000000005 10000000000000004.440892
10000000000000006 10000000000000006.661338
10000000000000007 10000000000000008.881784
10000000000000008 10000000000000008.881784
10000000000000009 10000000000000008.881784
>>> 
```
We have to go down to 1e9 to exercise the modified codepath in #8901.  And we have to increase resolution to see the benefit.  This is unmodified:
```
>>> d = int(1e9)
>>> for i in range(-10, 10):
...     print(d + i, "{:.16f}".format(float(d + i)))
... 
999999990 999999990.0000000607747097
999999991 999999990.9999999658793968
999999992 999999992.0000000486197678
999999993 999999992.9999999537244548
999999994 999999994.0000000364648258
999999995 999999994.9999999415695129
999999996 999999996.0000000243098839
999999997 999999997.0000001070502549
999999998 999999998.0000000121549419
999999999 999999999.0000000948953129
1000000000 1000000000.0000000000000000
1000000001 1000000001.0000000827403710
1000000002 1000000001.9999999434361371
1000000003 1000000003.0000000261765081
1000000004 1000000004.0000001089168791
1000000005 1000000005.0000001916572501
1000000006 1000000006.0000000523530161
1000000007 1000000007.0000001350933871
1000000008 1000000007.9999999957891532
1000000009 1000000009.0000000785295242
>>> 
```
This is with #8901:
```
>>> d = int(1e9)
>>> for i in range(-10, 10):
...     print(d + i, "{:.16f}".format(float(d + i)))
... 
999999990 999999990.0000000000000000
999999991 999999991.0000000000000000
999999992 999999992.0000000000000000
999999993 999999993.0000000000000000
999999994 999999994.0000000000000000
999999995 999999995.0000000000000000
999999996 999999996.0000000000000000
999999997 999999997.0000000000000000
999999998 999999998.0000000000000000
999999999 999999999.0000000000000000
1000000000 1000000000.0000000000000000
1000000001 1000000001.0000000000000000
1000000002 1000000002.0000000000000000
1000000003 1000000003.0000000000000000
1000000004 1000000004.0000000000000000
1000000005 1000000005.0000000000000000
1000000006 1000000006.0000000000000000
1000000007 1000000007.0000000000000000
1000000008 1000000008.0000000000000000
1000000009 1000000009.0000000000000000
>>> 
```
The current patch (#8905) formats the same as #8901, since the difference between adjacent ints here is much larger than the 2eps margin being used - it doesn't affect ints until we get to 1e16.

dpwe commented:
You can see the effect in decimals, though.

Current master:
```
>>> print("{:.16f}".format(1e9 - 0.000001))
999999999.9999991118215803
>>> print("{:.16f}".format(1e9 - 0.0000001))
999999999.9999998223643161
```
#8901 :
```
>>> print("{:.16f}".format(1e9 - 0.000001))
999999999.9999990463256836
>>> print("{:.16f}".format(1e9 - 0.0000001))
999999999.9999998807907104
```
#8905 :
```
>>> print("{:.16f}".format(1e9 - 0.000001))
999999999.9999990463256836
>>> print("{:.16f}".format(1e9 - 0.0000001))
1000000000.0000000000000000
```
#8905 is just a little eager to round up to a power of 10.

dpgeorge commented:
OK, thanks for the details.

So all the algorithms have problems (master, #8901, #8905 - this PR).  And that's fair enough, because we are trying to do this in a "short cut" way that doesn't take up too much code space.

But am I right to say that master and #8901 only get it wrong by at most a relative error of 2e-16, while this PR can get a relative error of up to 4e-16?  That's for `9999999999999995` rendering as `10000000000000000.000000`.

It would be nice if they all had the same maximum relative error, but this PR got it right "more of the time".  And also got whole numbers correct.  Or is that too much to ask?  Again, we could just write the full formatting routine that gets it perfect but that would take up too much code space.

----

Code size diff for the latest version of this PR is:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   +32 +0.005% 
unix nanbox:   +80 +0.014% 
      stm32:   +52 +0.013% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:  +100 +0.014% GENERIC
      esp32:  -132 -0.009% GENERIC
     mimxrt:   +56 +0.016% TEENSY40
 renesas-ra:   +48 +0.008% RA6M2_EK
        nrf:   +40 +0.022% pca10040
        rp2:   +32 +0.006% PICO
       samd:   +52 +0.037% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

dpwe commented:
The main purpose of this patch is to avoid printing anything but zeros after the decimal place when the number being printed is actually an integer (represented in a floating-point data structure).  Current master does not accomplish this, which leads to the perennial bug reports (incorrectly) questioning the underlying float math, e.g.
```
>>> print("{:.16f}".format(99))
99.0000000000000036
```
This patch fixes that problem by replacing the algorithm that repeatedly scaled the input by 0.1 (which cannot be exactly represented in binary, so led to cumulative errors) with an approach that repeatedly tests a comparison value of 10^X, which can be exactly represented up to 10^22.  Beyond that, you can't represent 10^X exactly in float64, so what to print becomes debatable.

However, differences between how the parser chose to represent, e.g., 1e33, and how `formatfloat` chose to render it under the new algorithm led to bad outcomes like:
```
>>> print(1e33)
9.999999999999998e+32
```
This arises because the reference values for 10^X in `formatfloat` are constructed on-the-fly from a table of 10^(2^Y) values, which do not exactly match the (accurate) values from the scanning code.  In some cases `formatfloat`'s reference values are slightly larger, so `formatfloat` tries to format at the next-smaller power of 10.  Even in most of these cases, final rounding corrects the result, but for 1e33, with the default resolution of 16 digits, the value does not round up.

My solution was reduce *every* 10^X comparison value by 2eps (2 steps in the least-significant value of the mantissa).  This was sufficient to avoid the original mistaken choice of power-of-10 for all positive powers of 10 in the unix build.  The consequence was that, for some non-integer values just below a power of 10, the value rounds up more aggressively.  Note that this *only* applies to values that are within 2eps of a power of ten, not for any other value.  In essence, this is a "goose" just to hide the floating point inaccuracies for these large values; I would argue the differences are immaterial, but they happen to be somewhat conspicuous.

Although this approach deliberately introduces deviations of 2eps in a few cases, the current algorithm has worst-case accuracy at least this bad because of the impact of repeated multiplication.  E.g. (on current master):
```
>>> print(6e63)
6.000000000000003e+63
```
That's also a relative error of 5e-16.  (I chose to look for values around 1e63, because the number of multiplications that the input is subject to depends on the number of bits set in the power of 10.)  The current patch prints this value exactly.

An alternative solution would be to store a table of 10^X for all X, or at least for a larger set of values such that the inaccurate comparison values were avoided.  This would indeed significantly increase object size (worst case increasing the table from 9 x 8 byte constants to around 300 x 8 byte constants).

There might be an algorithm to construct 10^X more exactly without a larger table but more slowly (e.g. by repeated multiplication by 10, which is exactly represented).  I tried a crude version of this; it worked up to 1e54, but was worse beyond that.

A less drastic, but less elegant, solution would be to special-case the particular values of 10^X where this formatting was a problem (10^X for X = 33, 34, 45, 46, 55, 59-63, ...) to make sure they formatted correctly.

Overall, I feel that the current approach of "aggressive rounding up for 10^X - 2eps" is a good solution - it's easily understood, and it's impact is limited only to this tiny subset of numbers just below 10^X.

For what it's worth, CPython does do a better job here.  E.g. for a large but inexact power of 10, this code will make it look as if it got it exactly right:
```
>>> print("{:.2f}".format(1e23))
100000000000000000000000.00
```
.. whereas CPython tells you the actual integer value of the float64 representation:
```
>>> print("{:.2f}".format(1e23))
99999999999999991611392.00
```
We'd need a totally different approach to match this.

dpgeorge commented:
OK, that all sounds reasonable.  It's good that this PR fixes the main cases that we keep seeing bug reports about, and fixes it in a relatively small amount of code (small code size increase).  As we know, to fix it fully requires a big code-size increase, so this is a good compromise.


> Although this approach deliberately introduces deviations of 2eps in a few cases, the current algorithm has worst-case accuracy at least this bad because of the impact of repeated multiplication.

Right, thanks for pointing that out.  In that case this PR doesn't make things any worse in terms of worst-case accuracy.

dpgeorge commented on py/formatfloat.c:
```diff
@@ -96,7 +96,21 @@ static inline int fp_isless1(float x) {
+union floatbits {
+    double f;
+    uint64_t u;
+};
```
I think you can use the existing `mp_float_union_t` instead of needed to add this union definition.

dpwe replied:
I could, but then I'd have to have different implementations of ```fp_ge_eps`` for float32 and float64 conditions.  I think it's more legible to have parallel types and a a single ``fp_ge_eps``.  Rewriting the float32 special-case functions for an unsigned int member is awkward.

dpgeorge replied:
Oh, I didn't realise the `MICROPY_FLOAT_IMPL_FLOAT` case just above also defines `union floatbits`.  In that case `mp_float_union_t` is exactly what you need, it's defined in `py/misc.h` for both the float and double cases.

dpgeorge commented:
When merging I will squash the commits together (unless you want to do that yourself and force-push to this branch).

But I see the commit "py/formatfloat: Fixed bug preventing rounding when %f promoted to %e." looks like it might be fixing an independent issue, in which case it should be kept as a separate commit.  Can you confirm this is fixing something else?  Do you have a test case for what this is fixing?  If so it might be good to add that test case as well.

dpwe commented:
The "bug preventing rounding" is exercised by this example (executing under master):
```
>>> print('{:f}'.format(8.888888888888888888e32))
8.888888e+32
>>> print('{:e}'.format(8.888888888888888888e32))
8.888889e+32
```
where the first case ('f' specifier) is incorrect.  The code originally skipped rounding if the original format specifier was 'f' and the absolute power of 10 is greater than the number of digits.  In the first case, the format has been promoted to 'e', so rounding is required.

However, this test appears to be unneeded under the new algorithm, where ```d``` is always the value of the next digit to be printed regardless of scaling.  I've removed it and verified that the test that exercised the original clause (e.g. ```print("%.1f" % (0.0006))```) does work correctly.

As such, the fixing of this bug is a side-effect of the other changes in this PR.  I don't think it needs a separate patch, do you?  I'll add a test to ```tests/float/float_format.py``` to guard against regressions.

dpwe commented:
The test is not pretty because the bug only appears in a situation where MicroPython and CPython already differ: Micropython will switch 'f' to 'e' when the fixed-point string would be too large for the buffer, but CPython always respects fixed-point formatting.

dpwe commented:
Hmm, I'll try a different test approach later.

dpgeorge commented:
> As such, the fixing of this bug is a side-effect of the other changes in this PR. I don't think it needs a separate patch, do you? I'll add a test to `tests/float/float_format.py` to guard against regressions.

If fixing is now a side-effect of the other changes, then it doesn't need a separate patch.  But, yes, please add a test to guard against regressions.

> The test is not pretty because the bug only appears in a situation where MicroPython and CPython already differ

In that case the test can go in a separate test file with an accompanying `.py.exp` file, with the expected output (then CPython won't be run, it'll just use the expected output to compare against).

dpwe commented:
Great.

I rebased to master, changed to using ```mp_float_union_t```, and added an expected output ```tests/float/float_format_ftoe.py.exp```.  Tests are passing, I think this is ready to go.

dpgeorge commented:
Running the new tests on bare-metal targets shows a few issues:

- on stm32 (PYBv1.0) and esp32 (both single precision float) the new `float_format_ftoe.py` test doesn't pass, it outputs: `8.888890e+32` instead of `8.888889e+32`.  These MCUs have hardware single precision float.

- on rp2 (PICO) `"{:.7g}".format(1e31)` gives `'9.999999e+30'` instead of `'1e31'`.  This MCU uses software routines for the single precision float.

Otherwise, the rest of the tests pass on these two boards (and they all pass on PYBD_SF6, which is stm32 with hardware double precision floats).

We need to get these tests passing on these bare-metal targets.  For the first case could it use `"%.5f"` instead of `"%f"`, or would that test something different?  For the second case can we just range up to 1e30 instead of 1e31?

dpgeorge commented:
New size diff is slightly better than before:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   +32 +0.005% 
unix nanbox:   +64 +0.011% 
      stm32:   +36 +0.009% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:   +36 +0.005% GENERIC
      esp32:  -156 -0.010% GENERIC
     mimxrt:   +40 +0.011% TEENSY40
 renesas-ra:   +32 +0.005% RA6M2_EK
        nrf:   +16 +0.009% pca10040
        rp2:    +0 +0.000% PICO
       samd:   +36 +0.026% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

dpwe commented:
* I changed ```tests/float_format_ftoe.py``` to use ```%.2f``` which I verified still exposes the original bug.
* I limited the range of printing powers of ten in ```tests/float_format_ints.py``` to stop at 10^30.

dpgeorge commented:
Thanks for fixing the tests, I checked on rp2 and stm32 and they are indeed now fixed.

Rebased and merged in f9cbe6bc47dd4f5b8e85178caecd6f0de22b4c34

This is a really nice improvement to have, it will definitely be appreciated by users.  And a bonus that the final code size increase was minimal.  Thanks very much @dpwe !

dpwe commented:
I'm not sure how passionate we feel about code size vs. execution speed.  I've continued playing with this (after looking at ```parsenum.c``` to see why we were experiencing the off-by-2eps problem) and I can get rid of the lookup tables altogether, saving ~144 bytes in float64 builds, at the cost of multiple calls to ```pow(10, i)```, which I expect is a lot slower.  That removes the need for ```fp_ge_eps```, but does expose problems with the parsing of some decimals (see my comment on #5832, as specifically revealed in ```tests/float/float_parse.py```).

dpwe commented:
I have a ```pow(10, i)```-based approach, along with a fix to ```py/parsenum.c``` to ignore trailing zeros in float value literals, implemented in: https://github.com/micropython/micropython/compare/master...dpwe:micropython:format_float_with_pow .  It's able to directly predict the base-10 exponent from the base-2 exponent in ```mp_float_union_t```, so it doesn't make too many calls to ```pow()```.  It eliminates the aggressive-rounding behavior, since the 10^X values in ```formatfloat``` directly match the ones used in ```parsenum```.

dpgeorge commented:
If you want to make a PR for that change, then we can evaluate it.
